xdc-plugin/xdc.cc

```diff
@@ -0,0 +1,242 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2012  Clifford Wolf <clifford@clifford.at>
+ *  Copyright (C) 2019  The Symbiflow Authors
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ *  ---
+ *
+ *   XDC commands
+ *
+ *   This plugin operates on the existing design and modifies its structure
+ *   based on the content of the XDC (Xilinx Design Constraints) file.
+ *   Since the XDC file consists of Tcl commands it is read using Yosys's
+ *   Tcl interpreter and processed by the new XDC commands imported to the
+ *   Tcl interpreter.
+ */
+#include <cassert>
+#include "kernel/register.h"
+#include "kernel/rtlil.h"
+#include "kernel/log.h"
+#include "libs/json11/json11.hpp"
+#include "../bank_tiles.h"
+
+USING_YOSYS_NAMESPACE
+
+PRIVATE_NAMESPACE_BEGIN
+
+enum SetPropertyOptions { INTERNAL_VREF };
+
+std::unordered_map<std::string, SetPropertyOptions> set_property_options_map  = {
+	{"INTERNAL_VREF", INTERNAL_VREF}
+};
+
+void register_in_tcl_interpreter(const std::string& command) {
+	Tcl_Interp* interp = yosys_get_tcl_interp();
+	std::string tcl_script = stringf("proc %s args { return [yosys %s {*}$args] }", command.c_str(), command.c_str());
+	Tcl_Eval(interp, tcl_script.c_str());
+}
+
+struct ReadXdc : public Frontend {
+	ReadXdc() : Frontend("xdc", "Read XDC file") {}
+
+	void help() YS_OVERRIDE {
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    read_xdc -part_json <part_json_filename> <filename>\n");
+		log("\n");
+		log("Read XDC file.\n");
+		log("\n");
+	}
+
+	void execute(std::istream *&f, std::string filename, std::vector<std::string> args, RTLIL::Design*) YS_OVERRIDE {
+                if (args.size() < 2) {
+                        log_cmd_error("Missing script file.\n");
+		}
+                Tcl_Interp *interp = yosys_get_tcl_interp();
+		size_t argidx = 1;
+		if (args[argidx] == "-part_json" && argidx + 1 < args.size()) {
+			bank_tiles = get_bank_tiles(args[++argidx]);
```

Comment:
This use of a global variable is not acceptable.  Can you propose other options?

Reply:
Here is an example of what you could do.

Have `ReadXdc` instance `GetPorts`, `GetIOBanks`, etc.  ReadXdc could own the `bank_tiles` instance, and pass a `std::function<const BankTiles&()>  to the constructors of `GetPorts`, `GetIOBanks` etc.  Upon destruction of `ReadXdc`, all passes and Frontend would do unregister from Yosys.

Reply:
Done